File: app/src/main/java/com/bassamalim/athkar/receivers/DailyUpdateReceiver.java
package com.bassamalim.athkar.receivers;

import android.Manifest;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.location.Location;
import android.util.Log;

import androidx.core.app.ActivityCompat;
import androidx.preference.PreferenceManager;

import com.bassamalim.athkar.Alarms;
import com.bassamalim.athkar.Constants;
import com.bassamalim.athkar.Keeper;
import com.bassamalim.athkar.PrayTimes;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationServices;

import java.util.Calendar;
import java.util.Date;
import java.util.TimeZone;

public class DailyUpdateReceiver extends BroadcastReceiver {

    private Context context;
    private SharedPreferences pref;
    private int time;

    @Override
    public void onReceive(Context gContext, Intent intent) {
        Log.i(Constants.TAG, "in daily update receiver");
        context = gContext;
        pref = PreferenceManager.getDefaultSharedPreferences(context);

        if (intent.getAction().equals("daily")) {
            time = intent.getIntExtra("time", 0);
            if (needed())
                process();
            else
                Log.i(Constants.TAG, "dead intent walking in daily update receiver");
        }
        else if (intent.getAction().equals("boot"))
            process();
    }

    private boolean needed() {
        int day = pref.getInt("last_day", 0);

        Calendar supposed = Calendar.getInstance();
        supposed.set(Calendar.HOUR_OF_DAY, time);

        Calendar now = Calendar.getInstance();

        return day != now.get(Calendar.DAY_OF_MONTH) && time >= now.get(Calendar.HOUR_OF_DAY);
    }

    private void process() {
        FusedLocationProviderClient fusedLocationClient =
                LocationServices.getFusedLocationProviderClient(context);

        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION)
                == PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(context,
                Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED) {

            fusedLocationClient.getLastLocation().addOnSuccessListener(this::update);
        }
    }

    private void update(Location location) {
        if (location != null) {
            new Keeper(context, location);

            Calendar[] times = getTimes(location);

            new Alarms(context, times);

            updateWidget();

            updated();
        }
        else
            Log.e(Constants.TAG, "Location is null in DailyUpdateService");
    }

    private Calendar[] getTimes(Location loc) {
        Calendar calendar = Calendar.getInstance();
        Date date = new Date();
        calendar.setTime(date);

        TimeZone timeZoneObj = TimeZone.getDefault();
        long millis = timeZoneObj.getOffset(date.getTime());
        double timezone = millis / 3600000.0;

        return new PrayTimes().getPrayerTimesArray(calendar, loc.getLatitude(),
                loc.getLongitude(), timezone);
    }

    private void updateWidget() {

    }

    private void updated() {
        Calendar today = Calendar.getInstance();

        SharedPreferences.Editor editor = pref.edit();
        editor.putInt("last_day", today.get(Calendar.DAY_OF_MONTH));
        editor.apply();
    }

}

File: app/src/main/java/bassamalim/hidaya/receivers/DailyUpdateReceiver.kt
package bassamalim.hidaya.receivers

import android.Manifest
import android.app.AlarmManager
import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.content.*
import android.content.pm.PackageManager
import android.location.Location
import android.util.Log
import androidx.core.app.ActivityCompat
import bassamalim.hidaya.Prefs
import bassamalim.hidaya.enums.LocationType
import bassamalim.hidaya.helpers.Alarms
import bassamalim.hidaya.other.Global
import bassamalim.hidaya.other.PrayersWidget
import bassamalim.hidaya.utils.DBUtils
import bassamalim.hidaya.utils.LocUtils
import bassamalim.hidaya.utils.PTUtils
import bassamalim.hidaya.utils.PrefUtils
import com.google.android.gms.location.LocationServices
import java.util.*

class DailyUpdateReceiver : BroadcastReceiver() {

    private lateinit var context: Context
    private lateinit var pref: SharedPreferences
    private var now = Calendar.getInstance()

    override fun onReceive(gContext: Context, intent: Intent) {
        Log.i(Global.TAG, "in DailyUpdateReceiver")

        context = gContext
        pref = PrefUtils.getPreferences(context)

        if ((intent.action == "daily" && needed()) || intent.action == "boot") {
            when (LocationType.valueOf(PrefUtils.getString(pref, Prefs.LocationType))) {
                LocationType.Auto -> locate()
                LocationType.Manual -> {
                    val cityId = PrefUtils.getInt(pref, Prefs.CityID)
                    if (cityId == -1) return
                    val city = DBUtils.getDB(context).cityDao().getCity(cityId)

                    val location = Location("")
                    location.latitude = city.latitude
                    location.longitude = city.longitude

                    update(location)
                }
                LocationType.None -> return
            }

            pickWerd()
        }
        else Log.i(Global.TAG, "dead intent walking in daily update receiver")

        setTomorrow()
    }

    private fun needed(): Boolean {
        return PrefUtils.getInt(pref, Prefs.LastDailyUpdateDay) != now[Calendar.DATE]
    }

    private fun locate() {
        if (ActivityCompat.checkSelfPermission(
                context, Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
            && ActivityCompat.checkSelfPermission(
                context, Manifest.permission.ACCESS_COARSE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
        ) {

            LocationServices.getFusedLocationProviderClient(context)
                .lastLocation.addOnSuccessListener {
                    location: Location? -> update(location)
            }
        }
    }

    private fun update(location: Location?) {
        if (location == null) {
            val storedLoc = LocUtils.retrieveLocation(pref)
            if (storedLoc == null) {
                Log.e(Global.TAG, "No available location in DailyUpdate")
                return
            }
        }
        else LocUtils.storeLocation(pref, location)

        val times = PTUtils.getTimes(pref, DBUtils.getDB(context))!!

        Alarms(context, times)

        updateWidget(times)

        updated()
    }

    private fun updateWidget(times: Array<Calendar?>) {
        val intent = Intent(context, PrayersWidget::class.java)
        intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
        intent.putExtra("times", times)

        val ids = AppWidgetManager.getInstance(context.applicationContext).getAppWidgetIds(
            ComponentName(context.applicationContext, PrayersWidget::class.java)
        )

        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)

        context.sendBroadcast(intent)
    }

    private fun updated() {
        val str = "Last Daily Update: ${now[Calendar.YEAR]}/${now[Calendar.MONTH] + 1}" +
            "/${now[Calendar.DATE]}" + " ${now[Calendar.HOUR_OF_DAY]}:${now[Calendar.MINUTE]}"

        pref.edit()
            .putInt("last_day", now[Calendar.DATE])
            .putString("last_daily_update", str)
            .apply()
    }

    private fun pickWerd() {
        pref.edit()
            .putInt("today_werd_page", Random().nextInt(Global.QURAN_PAGES) - 1)
            .putBoolean("werd_done", false)
            .apply()
    }

    private fun setTomorrow() {
        val intent = Intent(context.applicationContext, DailyUpdateReceiver::class.java)
        intent.action = "daily"

        val time = Calendar.getInstance()
        time[Calendar.DATE]++
        time[Calendar.HOUR_OF_DAY] = Global.DAILY_UPDATE_HOUR
        time[Calendar.MINUTE] = Global.DAILY_UPDATE_MINUTE

        val pendIntent = PendingIntent.getBroadcast(
            context.applicationContext, 1210, intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        val alarm =
            context.applicationContext.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        alarm.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, time.timeInMillis, pendIntent)
    }

}